### ISSUE TITLE
terra::writeCDF() uses significantly more memory for writing NetCDF files than terra::writeRaster()

### ISSUE THREAD

**Issue by pvjeetze** [rspatial/terra#1423]
We just came across some unexpected behavior where `terra::writeCDF` uses almost double the amount of memory for writing NetCDF files, as compared to `terra::writeRaster`, which is discouraged using and only provides limited functionality for writing `.nc` files. For us this is relevant because we frequently run SLURM jobs on a cluster with resource specifications fixed beforehand (CPUs, memory etc.). We find that the writing process becomes the memory bottleneck, even leading to unexpected out-of-memory kills in some cases. 

Example:

Max. memory load on the cluster with `writeCDF()`

![image](https://github.com/rspatial/terra/assets/50408549/cb5e91cd-8640-4ada-9362-a4ecd2cc97b8)

Max. memory load with `writeRaster()`

![image](https://github.com/rspatial/terra/assets/50408549/0935faa4-bc26-45ce-b862-8c28fa084a36)

**Comment by rhijmans** [comment]
I have made writeCDF more conservative. I assume this is because it needs to make an additional copy relative to what writeRaster does. In the future writeCDF should also go to the C++ side and then not suffer from this. 

A work-around for this problem would be to put a stronger limit on memory use via `terraOptions()`

**Comment by orhid** [comment]
i think this issue persists, i have not tested writeRaster, but writeCDF itself uses an insane amount of memory. i have a quick script which combines a couple of rasters using the `c` function. calling `mem_info` on the resulting object gives this output, which i think means that the object weighs about fourty gigs in memory

```
------------------------
Memory (GB) 
------------------------
check threshold : 1 (memmin)
available       : 682.83
allowed (50%)   : 341.42
needed (n=1)    : 46.26
------------------------
proc in memory  : TRUE
nr chunks       : 1
------------------------
```

the resulting file weighs a little more then twenty gigs (actually it looks like it weighs almost exactly half of what the object in R does), but the process of saving the file requires a whooping `660GB` of RAM, even though the `mem_info` claims it should not use more than 340

```
ID               Partition       Decl._mem            Max._node_mem.   Mem._%_usage
--               ---------      ---------            --------------   ------------
11670749     plgrid-bigmem   720.0GiB   659.9GiB(x1001c1s4b0n2)          91.7%
```

that was without explicitly setting limits with `terraOptions`, by setting the following
```
terraOptions(memfrac = 0.1, memmin = 12, memmax = 324)
```
i get this output from `mem_info`

```
------------------------
Memory (GB) 
------------------------
check threshold : 12 (memmin)
available       : 324 (memmax)
allowed (10%)   : 32.4
needed (n=1)    : 46.26
------------------------
proc in memory  : FALSE
nr chunks       : 15
------------------------
```

however the process still uses the exact same amount of memory as reported by slurm.

this might some miscofiguration on my part, or maybe the os, but even if we disregard the fact that memory limits set by `terraOptions` are disregarded, this massive use of memory is definitely not good

**Comment by orhid** [comment]
i have also checked the writeRaster function and it uses *only* 157GB of RAM, it also takes significantly less time than the writeCDF function (about 13 minutes on my machine)
